File: Java/src/org/hpms/dbg/Dump.java
package org.hpms.dbg;

import java.nio.ByteBuffer;

public class Dump {

   private static void dumpLine( String hexa, String ascii ) {
      hexa += "                                                                          ";
      hexa = hexa.substring( 0, 6+3*8+2+3*8 );
      System.err.println( hexa + " - " + ascii );
   }

   /**
    * Print a traditional dump: hexadecimal address, 8 hexadecimal bytes, a
    * separator, 8 hexadecimal bytes, a separator, 16 ASCII characters and a
    * newline.
    * <p>Sample:<pre>
    * 0000: 44 49 53 54 43 52 55 44 - 00 00 00 01 00 00 00 73  - DISTCRUD.......s
    * 0010: 00 00 00 2A 00 00 00 01 - 00 00 00 01 00 00 00 0D  - ...*............
    * 0020: 52 65 63 74 61 6E 67 6C - 65 20 30 30 31 40 76 C0  - Rectangle.001@v.
    * </pre></p>
    * @param bytes the array of bytes
    * @param from  the starting index
    * @param to    the ending index
    */
   public static void dump( byte[] bytes, int from, int to ) {
      int    addr   = 0;
      String addStr = "";
      String hexa   = "";
      String ascii  = "";
      for( int i = from; i < to; ++i ) {
         if( addr % 16 == 0 ) {
            if( addr > 0 ) {
               dumpLine( addStr + hexa, ascii );
               hexa  = "";
               ascii = "";
            }
            addStr = String.format( "%04X:", addr );
         }
         else if( addr % 8 == 0 ) {
            hexa = hexa + " -";
         }
         ++addr;
         final byte c = bytes[i];
         hexa = hexa + String.format( " %02X", c );
         ascii += ( c > 32 && c < 127 ) ? (char)c : '.';
      }
      if( ! hexa.isEmpty()) {
         dumpLine( addStr + hexa, ascii );
      }
   }

   /**
    * Dump an array of bytes, from 0 to length.
    * @param bytes the array of bytes.
    * @see {@link Dump#dump(byte[], int, int)}
    */
   public static void dump( byte[] bytes ) {
      dump( bytes, 0, bytes.length );
   }

   /**
    * Dump the remaining of a frame, from {@link java.nio.ByteBuffer#position()} to
    * {@link java.nio.ByteBufferByteBuffer#limit()}.
    * <p>Note: {@link java.nio.ByteBuffer#position()},
    * {@link java.nio.ByteBuffer#limit()} and
    * {@link java.nio.ByteBuffer#mark()} are not affected by the dump.</p>
    * @param frame the buffer to dump
    * @see {@link Dump#dump(byte[], int, int)}
    */
   public static void dump( ByteBuffer frame ) {
      dump( frame.array(), frame.position(), frame.limit());
   }
}

File: C/src/dbg/Dump.c
#include <dbg/Dump.h>
#include <string.h>

static void dumpLine( FILE * target, char * hexa, const char * ascii ) {
   hexa = strcat( hexa, "                                                                          " );
   hexa[6+( 3*8 )+2+( 3*8 )] = '\0';
   fprintf( target, "%s - %s\n", hexa, ascii );
}

void dbgDump_range( FILE * target, const byte * bytes, size_t from, size_t to ) {
   size_t addr        = 0;
   char   hexStr[200] = "";
   char   addStr[200] = "";
   char   hexa  [200] = "";
   char   ascii [200];
   size_t i;
   memset( hexStr, 0, sizeof( hexStr ));
   memset( addStr, 0, sizeof( addStr ));
   memset( hexa  , 0, sizeof( hexa   ));
   memset( ascii , 0, sizeof( ascii  ));
   for( i = from; i < to; ++i ) {
      byte c = bytes[i];
      if(( addr % 16 ) == 0 ) {
         if( addr > 0 ) {
            dumpLine( target, strcat( addStr, hexa ), ascii );
            memset( hexa , 0, sizeof( hexa  ));
            memset( ascii, 0, sizeof( ascii ));
         }
         sprintf( addStr, "%04X:", (unsigned int)addr );
      }
      else if(( addr % 8 ) == 0 ) {
         strcat( hexa, " -" );
      }
      ++addr;
      sprintf( hexStr, " %02X", c );
      strcat( hexa, hexStr );
      ascii[strlen(ascii)] = (( c > 32 ) && ( c < 127 )) ? (char)c : '.';
   }
   if( hexa[0] ) {
      dumpLine( target, strcat( addStr, hexa ), ascii );
   }
}

void dbgDump( FILE * target, const byte * bytes, size_t length ) {
   dbgDump_range( target, bytes, 0, length );
}
